packages/forms-web-app/src/controllers/common/enter-code.js

```diff
@@ -131,7 +201,97 @@ const postEnterCode = (views) => {
 	};
 };
 
+const getEnterCodeLPA = (views) => {
+	return async (req, res) => {
+		const {
+			body: { errors = {} },
+			params: { id }
+		} = req;
+		if (Object.keys(errors).length > 0) {
+			res.render(views.ENTER_CODE, {
+				errors: errors,
+				errorSummary: [{ text: errors.id.msg, href: '' }],
+				requestNewCodeLink: `/${views.REQUEST_NEW_CODE}`
+			});
+		} else {
+			res.render(views.ENTER_CODE, {
+				requestNewCodeLink: `/${views.REQUEST_NEW_CODE}`,
+				lpaUserId: id
+			});
+		}
+		return;
+	};
+};
+
+/**
+ * Creates the user object within the session object of the request for the successfully logged in LPAUser
+ * @param {ExpressRequest} req
+ * @param {UserId} userId
+ */
+const createLPAUserSession = async (req, userId) => {
+	let user = await getUserById(userId);
+	req.session.lpaUser = user;
+	req.session.save((err) => {
+		logger.error('Error storing LPA User Session');
+		logger.error(err);
+	});
+};
+
+const postEnterCodeLPA = (views) => {
+	return async (req, res) => {
+		const {
+			body: { errors = {}, errorSummary = [] },
+			params: { id }
+		} = req;
+
+		const emailCode = req.body['email-code'];
+
+		// if there are errors show error page
+		if (Object.keys(errors).length > 0) {
+			return renderErrorPageLPA(res, views.ENTER_CODE, {
+				lpaUserId: id,
+				emailCode,
+				errors,
+				errorSummary
+			});
+		}
+
+		if (isTestLPAAndEnvironment(emailCode, req.session)) {
+			try {
+				createLPAUserSession(req, id);
+				redirectToLPADashboard(res, views);
+				return;
+			} catch (e) {
+				logger.error(`Failed to create user session for user id ${id}`);
+				logger.error(e);
+				return {
+					valid: false,
+					action: enterCodeConfig.actions.lpaDashboard
+				};
+			}
+		}
+
+		// check token
+		let token = await isTokenValid(id, emailCode, req.session, id);
```

Comment:
Don't think the 4th param is used here, unless I've missed something

Reply:
Or the 3rd. I wonder if it read req.session.id at some point... Good spot.